stdlib, resources: Added pretty printing resource

Harshil2107 opened the pull request:
- Implemented a __str__  for AbstractResource __str__ prints resource category, id and version.
link to resources website is also printed.

Change-Id: Iad5825ff7d8d505ceb236e00dc49bb56055fc8f0

Harshil2107 committed:
stdlib, resources: Added pretty printing resource

- Implemented a __str__  for AbstractResource
__str__ prints resource category, id and version.
link to resources website is also printed.

Change-Id: Iad5825ff7d8d505ceb236e00dc49bb56055fc8f0

Harshil2107 committed:
stdlib, resources: Removed unused import

Change-Id: Iee54cc695c7c8ce146719ef583be424b792e2232

powerjg commented on src/python/gem5/resources/resource.py:
```diff
@@ -25,6 +25,7 @@
 # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 from abc import ABCMeta
+import json
```
Needed?

Harshil2107 replied:
The second commit removes the unneeded import, I will rebase everything into one commit after all requested changes are done.

powerjg commented on src/python/gem5/resources/resource.py:
```diff
@@ -741,6 +800,7 @@ def obtain_resource(
     # Once we know what AbstractResource subclass we are using, we create it.
     # The fields in the JSON object are assumed to map like-for-like to the
     # subclass contructor, so we can pass the resource_json map directly.
+
```
Extra newline?

Harshil2107 replied:
I removed this extra line in the second commit.

powerjg reviewed (commented):
lgtm. Two tiny things

BobbyRBruce commented on src/python/gem5/resources/resource.py:
```diff
@@ -91,12 +93,16 @@ def __init__(
         self._version = resource_version
 
+    def __str__(self):
+        return f"{self._category} resource({self._id}, {self._version})\nFor more information, please visit https://resources.gem5.org/resources/{self._id}?version={self._version}"
```
I don't think black cares about strings breaking line length rules so i typically do it manually.

```suggestion
        return f"{self._category} resource({self._id}, {self._version})\n"
                "For more information, please visit "
               f"https://resources.gem5.org/resources/{self._id}?"
               f"version{self._version}"
```

In python

```py
a_string = "hello this is an example of a string that may be a bit too long"
```

is the same as

```py
a_string = "hello "
           "this is an"
           " example of a string that may be "
           "a bi"
           "t too "
           "long"
```

BobbyRBruce commented on src/python/gem5/resources/resource.py:
```diff
@@ -122,6 +128,8 @@ class FileResource(AbstractResource):
         self,
         local_path: str,
+        id: Optional[str] = None,
+        category: Optional[str] = None,
```
Instead of having this as a parameter why not have the `AbstractResource` class enforce the creation of a `def get_category_name(cls) -> str` function. Then each of the categories has this implemented.

Harshil2107 committed:
stdlib: Addressed requested changes

- Added mulitline string for print message

- Added get_category_name method instead of having category as variable

Change-Id: I51e0e14a70e802453c21070711b200bc47994ba3

BobbyRBruce reviewed (approved):
Not sure if having `get_category_name` return the class name is my favorite choice here but it's fine for now and can be fixed later. Approving.
